[e2e demo 1/3] Add demo data generation & events

manolisliolios opened the pull request:
## Description 

Adds trading e2e example:
- Introduces changes to the SC (add events + change to using DOF for better lookups)
- Introduces all TS helpers to publish / create demo data.
- Introduces event indexer 
- Introduces API to read data

## Test Plan 

It works on my machine 😄 

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

amnn commented on examples/move/escrow/api/helpers/create-demo-escrows.ts:
```diff
@@ -0,0 +1,101 @@
+import { CONFIG } from '../config';
+import { getActiveAddress, getClient, signAndExecute } from '../sui-utils';
+
+const getOwnedBears = async () => {
```
Similar to the point I left elsewhere about making it clear that this app would work with any package, we should let people view all their owned objects and potentially lock any of them, not just the `DemoBear`.

manolisliolios replied:
The main motivation for all these demo data was to have some data to play with  the indexer + for the API to serve some data easily.

I accidently left the README file on the last PR in the stack #15832

amnn commented on examples/move/escrow/api/helpers/create-demo-data.ts:
How does this demo data get used?

manolisliolios replied:
It's part of the next PRs. 
Wanted to decouple them as the base functionality lives on the next ones.

amnn reviewed (commented):
Some initial questions/comments, mostly about the demo data generation.  The changes to the core contract look great, so if you wanted to go ahead and split that off, we can land that!

For the demo data, I agree that it would be good to have an easy way for people to generate objects to play with, but thoughts on the following?

- It would be nice to isolate it from the example, to make it clear that the example works with any objects, and we are just providing some demo data in another package for convenience. It also helps to see which parts of the example are for demo purposes only, and which parts are part of the implementation.
- The demo data just needs to generate objects -- we don't need to have the demo data lock any objects up-front, because that should be something that you can do in the UI, to make an object available for trading.

manolisliolios commented on examples/move/escrow/api/.gitignore:
I had the exact same feeling and forgot to raise that question on the description. Agreed + will move it!

manolisliolios commented on examples/move/escrow/api/helpers/cancel-escrow.ts:
```diff
@@ -0,0 +1,23 @@
+	await signAndExecute(txb, CONFIG.NETWORK);
+};
+
+cancelEscrow('0x4f56f3c3cfcf28448a0b4f19d1abe08927d84a677b25dc31b1d8d1dfe1888bcf');
```
It was just a sample run that I left as is. Could be removed completely, it just shows how to call it!

amnn replied:
Would this be called when you import this module? That's what seemed strange to me.

manolisliolios replied:
Yes, but in reality, this was just a callable file (I had not exported `cancelEscrow` at that point) so it was meant to be used only by calling it directly (as a script, calling `npx ts-node helpers/cancel-escrow.ts`).

I think I changed this on my latest PR though.

manolisliolios commented on examples/move/escrow/contract/sources/demo_bear.move:
I am not sure what kind of objects we could use that do pre-exist, that's the main reason I created a collection to demo an NFT-like object.

amnn replied:
Yeah, I think it's fair to offer a new package to be that demo type. The ideal case would be that we had some other app example that we could suggest people use to create an NFT or something, but having a separate demo package is fine for now!

manolisliolios commented on examples/move/escrow/contract/sources/lock.move:
```diff
@@ -31,31 +38,63 @@ module escrow::lock {
+    }
+
+    struct LockDestroyed has copy, drop {
+        /// The ID of the `Locker` object.
```
Hehe, the struct name was `Locked` so I followed it everywhere 😆

manolisliolios commented on examples/move/escrow/api/helpers/publish-contract.ts:
It becomes runnable, for instance by calling `npx ts-node helpers/publish-contract.ts` 

I could also make it a self invoking function, but that takes away re-usability from other files.

manolisliolios commented:
> It would be nice to isolate it from the example, to make it clear that the example works with any objects, and we are just providing some demo data in another package for convenience. It also helps to see which parts of the example are for demo purposes only, and which parts are part of the implementation.

Agreed + I became a bit lazy here (also didn't want to add too many packages, but we're moving this out to its own place, so it'll not look that bad). Will refactor this to publish the demo collection in a separate package.

> The demo data just needs to generate objects -- we don't need to have the demo data lock any objects up-front, because that should be something that you can do in the UI, to make an object available for trading.

Agreed, however we also have the indexer + API there and the main motivation for creating the demo data was to test these. It's actually not locking all of the objects, but a percentage of them. We might even move all of this functionality in test utils if we also add e2e tests.

manolisliolios commented:
@amnn I've updated the structure: decoupled demo objects contract + core escrow contract, updated demo scripts (to use the new config), generalized some utils (e..g publishing)

amnn commented:
> Agreed, however we also have the indexer + API there and the main motivation for creating the demo data was to test these. It's actually not locking all of the objects, but a percentage of them. We might even move all of this functionality in test utils if we also add e2e tests.

Got it, that makes a lot of sense!  As long as from the guide perspective, we're using this only in automated tests, and not part of the demo set-up for the running app, we're all good there.

amnn commented on examples/trading/api/config.ts:
cc @rvantonder -- this is a good example of how people are currently doing address management for packages. A useful reference when thinking about automated address management.

manolisliolios replied:
For local setups, that's true. In most mainnet/testnet deployments (prod), we usually use static values from ts files to make this even more predictable + safe (not relying on any env configs, unless they are committed to the repo).

Something like:
```typescript
const MAINNET_CONFIG =  {
    packageId: '0xMyMainnetAddress',
}

const TESTNET_CONFIG = {
   packageId: '0xMyTestnetAddress'
}

export const CONFIG = process.env.NETWORK === 'testnet' ? TESTNET_CONFIG : MAINNET_CONFIG
```

amnn commented on examples/trading/api/helpers/cancel-escrow.ts:
```diff
@@ -0,0 +1,22 @@
+import { CONFIG } from '../config';
+import { getActiveAddress, signAndExecute } from '../sui-utils';
+
+/// Demo PTB to cancel an escrow.
```
Out of interest, is `///` a valid doc comment designator for TS/JS?

manolisliolios replied:
Agh, no. Brain lag 😆 Let me correct them all, and thanks for flagging.

The (recommended?) way for TS is https://tsdoc.org/

```typescript
/**
 * A comment
 * @param test ...
 */
```

amnn commented on examples/trading/api/helpers/create-demo-data.ts:
```diff
@@ -0,0 +1,60 @@
+const createDemoLockedObjects = async (totalBears: number) => {
+	if (totalBears < 5) throw new Error('Please create at least 5 bears to run this script.');
+	const txb = new TransactionBlock();
+	const bears = [];
```
nit: `bears` is a little confusing as a name here, because it also ends up containing `Locked<DemoBear>`s and `Key`s as well -- can we tweak the name? How about: `toTransfer` or something?

amnn commented on examples/trading/contracts/demo/Move.toml:
```diff
@@ -0,0 +1,9 @@
+version = "0.0.1"
+
+[dependencies]
+Sui = { local = "../../../../crates/sui-framework/packages/sui-framework" }
```
Should we depend on the `git` version of `sui-framework`, so it's easy for people to copy/paste this? (What do we do for other E2E examples?)

manolisliolios replied:
@amnn it indeed becomes more usable that way, thanks!

amnn commented on examples/trading/contracts/escrow/sources/shared.move:
```diff
@@ -86,19 +99,25 @@ module escrow::shared {
+            escrow_id: object::uid_to_inner(&id),
+        });
+
         object::delete(id);
```
Just to jog my memory -- this means that this version is building on top of the version of this example from the PR where I adapt it for shared object deletion, right?

manolisliolios replied:
Yeah, this PR points to your branch which introduces the changes to test scenario too, adding support for shared object deletion.

amnn replied:
Ah, I hadn't spotted that, yep that makes sense!

amnn reviewed (approved):
LGTM! Thanks @manolisliolios
